imperial.js

```diff
@@ -3,23 +3,50 @@ const Action = require("./action");
 
 class Imperial {
   static fromLog(log) {
-    return new Imperial(log);
+    let game = new Imperial();
+    log.forEach((entry) => game.tick(entry));
+    return game;
   }
 
-  constructor(log) {
-    this.log = log;
+  constructor() {
+    this.log = [];
+    this.nations = this.setupNations();
+    this.players = {};
   }
 
   get state() {
     return {
       availableActions: this.availableActions(),
       investorCardHolder: this.investorCardHolder(),
-      nations: this.nations(),
-      players: this.players(),
+      nations: this.nationsState(),
+      players: this.playersState(),
       provinces: this.provinces(),
     };
   }
 
+  setupNations() {
+    let nations = {};
+    [Nation.AH, Nation.IT, Nation.FR, Nation.GB, Nation.GE, Nation.RU].forEach(
+      (nation) => (nations[nation] = { controller: "", treasury: 0 })
+    );
+    return nations;
+  }
+
+  tick(action) {
+    if (action.type === "playerSeating") {
+      this.order = action.payload.order;
+      action.payload.order.forEach(
+        (player) => (this.players[player] = { cash: 13 })
+      );
+    } else if (action.type === "assignStartingNation") {
+      this.nations[action.payload.nation].controller = action.payload.player;
+    } else if (action.type === "bondPurchase") {
+      this.nations[action.payload.nation].treasury += action.payload.cost;
+      this.players[action.payload.player].cash -= action.payload.cost;
+    }
+    this.log.push(action);
+  }
```

Comment:
If we change `this.nations` to be a Map, we'll need to change our usage to `this.nations.get(action.payload.nation)`.